homeassistant/components/pvpc_hourly_pricing/sensor.py

```diff
@@ -0,0 +1,195 @@
+            mins_update,
+        )
+        await self.async_update_prices(dt_util.utcnow())
+        self._init_done = True
+        await self.async_update_ha_state(True)
+
+    @property
+    def unique_id(self) -> Optional[str]:
+        """Return a unique ID."""
+        return self._unique_id
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._pvpc_data.state
+
+    @property
+    def available(self) -> bool:
+        """Return True if entity is available."""
+        return self._pvpc_data.state_available
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        return self._pvpc_data.attributes
+
+    async def async_update(self, *_args):
+        """Update the sensor state."""
+        if not self._init_done:
+            # abort until added_to_hass is finished
+            return
+
+        now = dt_util.utcnow()
+        if self._pvpc_data.process_state_and_attributes(now):
+            await self.async_update_ha_state()
+        else:
+            # If no prices present, download and schedule a future state update
+            self._pvpc_data.state_available = False
+            self.async_schedule_update_ha_state()
+
+            if self._pvpc_data.source_available:
+                _LOGGER.debug(
+                    "[%s]: Downloading prices as there are no valid ones",
+                    self.entity_id,
+                )
+                async_track_point_in_time(
+                    self.hass,
+                    self.async_update,
+                    now + timedelta(seconds=self._pvpc_data.timeout),
+                )
+            await self.async_update_prices(now)
+
+    async def async_update_prices(self, now):
+        """Update electricity prices from the ESIOS API."""
+        prices = await self._pvpc_data.async_update_prices(now)
+        if not prices and self._pvpc_data.source_available:
+            self._num_retries += 1
+            if self._num_retries > 2:
+                _LOGGER.warning(
+                    "Repeated bad data update, mark component as unavailable source"
+                )
+                self._pvpc_data.source_available = False
+                return
+
+            retry_delay = 3 * self._pvpc_data.timeout
+            _LOGGER.debug(
+                "Bad update[retry:%d], will try again in %d s",
+                self._num_retries,
+                retry_delay,
+            )
+            async_track_point_in_time(
+                self.hass,
+                self.async_update_prices,
+                dt_util.now() + timedelta(seconds=retry_delay),
+            )
+            return
+
+        if not prices:
+            _LOGGER.debug(
+                "Data source unavailable since %s",
+                self.hass.states.get(self.entity_id).last_changed,
```

Comment:
Can you call
`self.last_changed` instead?

Reply:
🤯

Reply:
It does not work :(

That is a property of the `State`, from the state machine. It is not available nor registered in the `Entity`. Now I think I remember trying it before using that :)

Reply:
I'm not sure if grabbing the state is ok to do here, but it seems like we shouldn't do that.  Probably need another reviewer who is more familiar with the accepted practices to be sure.

Looks like `_num_retries` is already being store, maybe store the last success as well?
```
self._last_update_success
```

Starting to think with _num_retries / _last_update_success this warrants another class to have better separation of concerns

Reply:
I've removed the usage of `last_updated`, as it was just for debug logs.